### PR TITLE
Ban users logic and backends

### PR DESCRIPTION
**Pull request by Ivanca**
2 backends: DatabaseBannedUsersBackend, UserprofileBannedUsersBackend but the later is not implemented yet.

image

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -16,6 +16,7 @@ factory-boy==2.10.0
 djangorestframework-jwt==1.11.0
 Sphinx==1.7.8
 sphinx-autobuild==0.7.1
--e git://github.com/edx/opaque-keys.git@f3942654aa9e341c3d1d4b459ee705a2c5a2dc73#egg=edx_opaque_keys
+-e git://github.com/edx/opaque-keys.git@8e0697778a13b8214ef1862e3d91eccc04af1663#egg=edx_opaque_keys
```
Comment (felipemontoya) [inline requirements.txt:19]: Is this the absolute last version?

File: seb_openedx/dashboard/urls.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+""" urls for dashboard """
+from django.conf.urls import url
+from django.contrib.admin.views.decorators import staff_member_required
+from seb_openedx.dashboard.views import TableView
+
+urlpatterns = [  # pylint: disable=invalid-name
+    url(r'^$', staff_member_required(TableView.as_view())),
```
Comment (felipemontoya) [inline seb_openedx/dashboard/urls.py:7]: This means the global is_staff flag on the django user, right?
Eventually we might want to do course_staff, but for the time being, I think this is good

File: seb_openedx/middleware.py [hunk 3]
```diff
@@ -12,9 +12,11 @@
 from seb_openedx.permissions import get_enabled_permission_classes
 from seb_openedx.seb_courseware_index import SebCoursewareIndex
 from seb_openedx.edxapp_wrapper.get_chapter_from_location import get_chapter_from_location
+from seb_openedx.user_banning import is_user_banned, ban_user
 
 SEB_WHITELIST_PATHS = getattr(settings, 'SEB_WHITELIST_PATHS', [])
 SEB_BLACKLIST_CHAPTERS = getattr(settings, 'SEB_BLACKLIST_CHAPTERS', [])
+BANNING_ENABLED = getattr(settings, 'SEB_USER_BANNING_ENABLED', False)
```
Comment (felipemontoya) [inline seb_openedx/middleware.py]: Why is this defaulting to False, I would asume this should be always on

File: seb_openedx/dashboard/views.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
+"""The generic views for the exc-core plugin project"""
+
+from __future__ import unicode_literals
+import django_tables2 as tables
```
Comment (felipemontoya) [inline seb_openedx/dashboard/views.py]: What do we get from this django_tables2 package? for this to be installed in production we must go out of the regular way, so this requirement better be super worth it

File: seb_openedx/middleware.py [hunk 5]
```diff
@@ -38,24 +39,34 @@ def process_view(self, request, view_func, view_args, view_kwargs):
...
+                access_denied = True
+
+            if access_denied:
+                return self.handle_access_denied(request, view_func, view_args, view_kwargs, course_key, user_name, banned)
 
         return None
 
     # pylint: disable=too-many-arguments
-    def handle_access_denied(self, request, view_func, view_args, view_kwargs, course_key):
+    def handle_access_denied(self, request, view_func, view_args, view_kwargs, course_key, user_name, banned):
```
Comment (felipemontoya) [inline seb_openedx/middleware.py]: This does look like we are starting to put too many arguments. Do we really want to handle banning inside of this function?

File: seb_openedx/middleware.py [hunk 6]
```diff
@@ -105,25 +116,25 @@ def is_blacklisted_chapter(self, request, course_key):
                     return True
         return False
 
-    def courseware_error_response(self, request, *view_args, **view_kwargs):
+    def courseware_error_response(self, context, request, *view_args, **view_kwargs):
```
Comment (felipemontoya) [inline seb_openedx/middleware.py:124]: +1

File: seb_openedx/models.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+""" Models """
+from __future__ import absolute_import, unicode_literals
+from django.db import models
+from opaque_keys.edx.django.models import CourseKeyField
+
+
+class ForbiddenCourseAccess(models.Model):
+    """ user_id <-> course_id relation to mark forbidden users """
+    username = models.CharField(max_length=150)  # from django.contri.auth.user
```
Comment (felipemontoya) [inline seb_openedx/models.py]: In theory a user should not change its username, so this is not so problematic, but I still think the user_id is safer

File: seb_openedx/templates/seb-manage-users/table.html [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        <div class="col-10">
+          <br>
+          <h1 class="mt-2">SEB - User Admin</h1>
+          <br>
+          {% render_table table %}
+        </div>
+      </div>
+    </div>
+  </body>
+</html>
```
Comment (felipemontoya) [inline seb_openedx/templates/seb-manage-users/table.html]: Emtpy line

File: seb_openedx/user_banning.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    @abc.abstractmethod
+    def unban_user(self, username, course_key, last_modified_time, unbanned_by):
+        """ is_user_banned abstract method """
+        pass
+
+
+class DatabaseBannedUsersBackend(BannedUsersBackend):
+    """ Database backend """
+    def is_user_banned(self, username, course_key):
+        """ is_user_banned implementation """
```
Comment (felipemontoya) [inline seb_openedx/user_banning.py]: This docstring could give more info. It is clearly a placeholder, but now it makes sense to actually write it

File: seb_openedx/user_banning.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        raise NotImplementedError()
+
+    def unban_user(self, username, course_key, last_modified_time, unbanned_by):
+        """ ban_user implementation """
+        raise NotImplementedError()
+
+
+def _get_back_end():
+    """ get back end util """
+    if getattr(settings, 'SEB_USER_BANNING_BACKEND', 'database') == 'database':
```
Comment (felipemontoya) [inline seb_openedx/user_banning.py]: to be more in line with django, the name of the backend should be 'seb_openedx.user_banning.DatabaseBannedUsersBackend' or 'DatabaseBannedUsersBackend' if we are already asuming the backend lives in this module and here the import would require no special case scenario for the `'database'` string.

**Review by felipemontoya — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 1 above)
Comment (Ivanca) [reply requirements.txt:19]: Yes -the one from Mar 26, 2018- otherwise it fails.

File: seb_openedx/dashboard/urls.py (hunk 2 above)
Comment (Ivanca) [reply seb_openedx/dashboard/urls.py:7]: Understood.

File: seb_openedx/middleware.py (hunk 3 above)
Comment (Ivanca) [reply seb_openedx/middleware.py]: Ok

File: seb_openedx/middleware.py (hunk 5 above)
Comment (Ivanca) [reply seb_openedx/middleware.py]: The view needs to know about the banning; how else than as an argument?

File: seb_openedx/models.py (hunk 7 above)
Comment (Ivanca) [reply seb_openedx/models.py]: user_id doesn't exist in the User model, it's that something by edx?

File: seb_openedx/templates/seb-manage-users/table.html (hunk 8 above)
Comment (Ivanca) [reply seb_openedx/templates/seb-manage-users/table.html]: Ok

File: seb_openedx/user_banning.py (hunk 9 above)
Comment (Ivanca) [reply seb_openedx/user_banning.py]: Ok

File: seb_openedx/user_banning.py (hunk 10 above)
Comment (Ivanca) [reply seb_openedx/user_banning.py]: Ok

File: seb_openedx/dashboard/views.py (hunk 4 above)
Comment (Ivanca) [reply seb_openedx/dashboard/views.py]: Ok; I'll hardcode the HTML instead then.

**Comment by Ivanca** [comment]
@felipemontoya commit addressing feedback https://github.com/eduNEXT/seb-openedx/pull/13/commits/724fa51a6e3ec178901b440981d873333611e512

File: seb_openedx/middleware.py (hunk 5 above)
Comment (felipemontoya) [reply seb_openedx/middleware.py]: We can make a function called handle_banning, which is called on line 51 and use a property, so at any point the methods can call `self.is_banned` o `self.access_denied`

File: seb_openedx/models.py (hunk 7 above)
Comment (felipemontoya) [reply seb_openedx/models.py]: at the user model is called user.id, but it is the same

File: seb_openedx/templates/seb-manage-users/table.html (hunk 8 above)
Comment (felipemontoya) [reply seb_openedx/templates/seb-manage-users/table.html]: still missing the emtpy line. Edit: not any more. I think github did not show the last changes

Comment (felipemontoya) [reply seb_openedx/templates/seb-manage-users/table.html]: I see it again. @github, this is confusing

File: seb_openedx/user_banning.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    def unban_user(self, username, course_key, last_modified_time, unbanned_by):
+        """ is_user_banned abstract method """
+        pass
+
+
+class DatabaseBannedUsersBackend(BannedUsersBackend):
+    """ Database backend """
+    def is_user_banned(self, username, course_key):
+        """ is_user_banned implementation """
+        return len(ForbiddenCourseAccess.objects.filter(username=username, course_id=course_key))  # pylint: disable=no-member
```
Comment (felipemontoya) [inline seb_openedx/user_banning.py]: Shouldn't this check only for objects that have `access_blocked` in `true`?

Comment (Ivanca) [reply seb_openedx/user_banning.py]: Fixed on last commit

File: seb_openedx/user_banning.py (hunk 9 above)
Comment (Ivanca) [reply seb_openedx/user_banning.py]: Fixed on last commit

File: seb_openedx/templates/seb-manage-users/table.html (hunk 8 above)
Comment (Ivanca) [reply seb_openedx/templates/seb-manage-users/table.html]: Fixed on last commit

File: seb_openedx/models.py (hunk 7 above)
Comment (Ivanca) [reply seb_openedx/models.py]: Ok

File: seb_openedx/middleware.py (hunk 5 above)
Comment (Ivanca) [reply seb_openedx/middleware.py]: Fixed on last commit (banned is now a instance property)

**Comment by Ivanca** [comment]
@felipemontoya Latest commit adressing feedback 950e1e3, review needed.

File: seb_openedx/models.py (hunk 7 above)
Comment (felipemontoya) [inline seb_openedx/models.py]:
```suggestion
    username = models.CharField(max_length=150)  # from django.contrib.auth.user
```

File: seb_openedx/templates/seb-403-error-message.html [hunk 12]
```diff
@@ -8,4 +8,7 @@ <h1><%block name="pageheader">${_("Error: Access not allowed")}</%block></h1>
     <%block name="pagecontent">
         ${Text(_(' This course can only be accessed with "Safe Exam Browser"'))}
```
Comment (felipemontoya) [inline seb_openedx/templates/seb-403-error-message.html]:
```suggestion
        ${Text(_('This course can only be accessed with "Safe Exam Browser"'))}
```

File: seb_openedx/templates/seb-403-error-message.html [hunk 13]
```diff
@@ -8,4 +8,7 @@ <h1><%block name="pageheader">${_("Error: Access not allowed")}</%block></h1>
...
     </%block>
+    % if banned:
+	${Text(_(' Additionally you have been banned, you may ask your teacher or systems administrator to unblock you.'))}
```
Comment (felipemontoya) [inline seb_openedx/templates/seb-403-error-message.html]:
```suggestion
	${Text(_('Additionally you have been banned, you may ask your teacher or systems administrator to unblock you.'))}
```

**Review by felipemontoya — APPROVED** [review]
I left a few suggestions inline. After that we should squash the noise and we are ready to merge

**Pull request merged**